### PR TITLE
Interactive SCCharts

### PR DESCRIPTION
**Pull request by soerendomroes**
See
https://github.com/eclipse/elk/pull/878
https://github.com/kieler/KLighD/pull/143
https://github.com/kieler/klighd-vscode/pull/121

- [ ] Add synthesis option to enable interactive layout for all nodes.

### CODE REVIEW

File: plugins/de.cau.cs.kieler.sccharts.ide/src/de/cau/cs/kieler/sccharts/ide/language/server/SCTXConstraintSerializer.xtend [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.List
+import java.util.Map
+import org.eclipse.elk.graph.properties.IProperty
+import org.eclipse.lsp4j.Position
+import org.eclipse.lsp4j.Range
+import org.eclipse.lsp4j.TextEdit
+
+/**
+ * @author sdo
+ *
```
Comment (Eddykasp) [inline plugins/de.cau.cs.kieler.sccharts.ide/src/de/cau/cs/kieler/sccharts/ide/language/server/SCTXConstraintSerializer.xtend:35]: document classes

File: plugins/de.cau.cs.kieler.sccharts.ui/src/de/cau/cs/kieler/sccharts/ui/synthesis/ControlflowRegionSynthesis.xtend [hunk 2]
```diff
@@ -77,6 +77,11 @@ class ControlflowRegionSynthesis extends SubSynthesis<ControlflowRegion, KNode>
     override performTranformation(ControlflowRegion region) {
         val node = region.createNode().associateWith(region);
         
+        // default value of interactive layout for regions
+        if (!region.annotations.contains(CoreOptions.INTERACTIVE_LAYOUT)) {
+            node.setLayoutOption(CoreOptions::INTERACTIVE_LAYOUT, true)
+        }
```
Comment (Eddykasp) [inline plugins/de.cau.cs.kieler.sccharts.ui/src/de/cau/cs/kieler/sccharts/ui/synthesis/ControlflowRegionSynthesis.xtend:83]: This confuses me, if the region does not have an interactive layout annotation then it should be set to true?

Comment (soerendomroes) [reply plugins/de.cau.cs.kieler.sccharts.ui/src/de/cau/cs/kieler/sccharts/ui/synthesis/ControlflowRegionSynthesis.xtend:83]: @Drakae Do you remember why you need this?

Comment (Drakae) [reply plugins/de.cau.cs.kieler.sccharts.ui/src/de/cau/cs/kieler/sccharts/ui/synthesis/ControlflowRegionSynthesis.xtend:83]: i think the idea was that the user must especially turn off interactive layout if it is not wanted

File: plugins/de.cau.cs.kieler.sccharts.ui/src/de/cau/cs/kieler/sccharts/ui/synthesis/SCChartsSynthesis.xtend [hunk 3]
```diff
@@ -351,13 +351,23 @@ class SCChartsSynthesis extends AbstractDiagramSynthesis<SCCharts> {
     override List<? extends IGraphElementVisitor> getAdditionalLayoutConfigs(KNode viewModel) {
         val List<IGraphElementVisitor> additionalLayoutRuns = newArrayList
         // Add interactive Layout run.
-        if ((!viewModel.getChildren().isEmpty() && viewModel.getChildren().get(0)
-                        .getProperty(CoreOptions.INTERACTIVE_LAYOUT))) {
+        if (!viewModel.getChildren().isEmpty() && (viewModel.getChildren().get(0)
+                        .getProperty(CoreOptions.INTERACTIVE_LAYOUT) || isChildInteractive(viewModel))) {
             additionalLayoutRuns.add(new CompoundGraphElementVisitor(
                     new InteractiveRectPackingGraphVisitor(),
                     new InteractiveLayeredGraphVisitor()));
```
Comment (soerendomroes) [inline plugins/de.cau.cs.kieler.sccharts.ui/src/de/cau/cs/kieler/sccharts/ui/synthesis/SCChartsSynthesis.xtend:358]: I propose to add an additional graph visitor here, which disables model order if we are in interactive mode.

File: plugins/de.cau.cs.kieler.sccharts.ui/src/de/cau/cs/kieler/sccharts/ui/synthesis/ControlflowRegionSynthesis.xtend (hunk 2 above)
Comment (soerendomroes) [reply plugins/de.cau.cs.kieler.sccharts.ui/src/de/cau/cs/kieler/sccharts/ui/synthesis/ControlflowRegionSynthesis.xtend:83]: Maybe this should be a synthesis property and activate it for all nodes if wanted.

**Pull request closed without merging**